### PR TITLE
[SUPERSEDED] Use ellipsis instead of silent truncation in stringOf

### LINKED ISSUE #12337: REPL element truncation should not silently truncate
## reproduction steps

using Scala 2.13.4,

```scala
scala -Dscala.repl.maxprintstring=10000
Welcome to Scala 2.13.4 (OpenJDK 64-Bit Server VM, Java 15).
Type in expressions for evaluation. Or try :help.

scala> Iterator.from(1).take(10000).to(List)
[snip] 998, 999, 1000)
```

## problem

The truncated collection is rendered without an ellipsis.

Reported at https://github.com/lampepfl/dotty/issues/11377

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes https://github.com/scala/bug/issues/12337

```
➜  ~ scala -Dscala.repl.maxprintstring=50
Welcome to Scala 2.13.10-20221005-211743-e04f32b (OpenJDK 64-Bit Server VM, Java 19).
Type in expressions for evaluation. Or try :help.

scala> (1 to 1000).map(i => "X"*5).toList
val res0: List[String] = List(XXXXX, XXXXX, XXXXX, XXXXX, XXXXX, ...)

scala> (1 to 1000).map(i => "X"*50).toList
val res1: List[String] = List(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...)

scala> "List(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...)".length
val res2: Int = 50

scala> "X"*50
val res3: String = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

scala> "X"*51
val res4: String = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX...
```

### CODE REVIEW

**Comment by som-snytt** [comment]
@SethTisue it's so pretty. I wonder if you'll feel sorry if 2.13.10 doesn't have it. I mean the rueful thoughts that occur to you years later while tarrying in a coffee shop during an unexpected rain storm.

File: src/library/scala/runtime/ScalaRunTime.scala [hunk 1]
```diff
@@ -227,51 +252,166 @@ object ScalaRunTime {
...
-      case x: Product if isTuple(x)     => x.productIterator.map(inner).mkString("(", ",", ")")
+      case x: collection.Map[_, _]      => stringify(x.iterator, x.collectionClassName, mapInner)
+      case x: Iterable[_]               => stringify(x.iterator, x.collectionClassName, inner)
+      case x: Product1[_] if isTuple(x) => s"(${inner(x._1)},)" // that special trailing comma
+      case x: Product if isTuple(x)     => x.productIterator.map(inner).mkString("(", (if (verboseProduct) ", " else ","), ")")
+      case x: Product if verboseProduct && x.productArity > 0 =>
+        if (x.productArity == 1 && !x.productElement(0).isInstanceOf[Boolean])
+          s"${x.productPrefix}(${inner(x.productElement(0))})"
+        else
+          _toString1(x, inner(_))
```
Comment (lrytz) [inline src/library/scala/runtime/ScalaRunTime.scala:385]: `_toString0` is unused, and this looks almost like it

File: src/library/scala/runtime/ScalaRunTime.scala [hunk 2]
```diff
@@ -171,20 +194,22 @@ object ScalaRunTime {
...
    *
-   * @param   arg   the value to stringify
-   * @return        a string representation of arg.
+   *  Does not handle null result of toString. (See replStringOf.)
+   *
+   *  @param   arg              the value to stringify
+   *  @param   maxLength        limit the length of the result string
+   *  @param   verboseProduct   extra user-friendly
+   *  @return  a string representation of arg.
    */
```
Comment (lrytz) [inline src/library/scala/runtime/ScalaRunTime.scala:209]: Nit, but the Scaladoc should be on the 3-param overload

**Review by lrytz — COMMENTED** [review]
The new pretty version is not yet in use, IIUC?

File: src/library/scala/runtime/ScalaRunTime.scala (hunk 2 above)
Comment (som-snytt) [reply src/library/scala/runtime/ScalaRunTime.scala:209]: I noticed that, but my thought was, scaladoc should be smart enough to figure out overloads. I just drove someone to the airport, and my ordinary Honda knows how to stay inside the white lines!

File: src/library/scala/runtime/ScalaRunTime.scala (hunk 1 above)
Comment (som-snytt) [reply src/library/scala/runtime/ScalaRunTime.scala:385]: Maybe not unused forever, or at least by me in _my_ REPL.

**Comment by som-snytt** [comment]
The API is used by Scala 3. I will PR there to make verbose opt-in somehow.

**Comment by SethTisue** [comment]
>  I mean the rueful thoughts that occur to you years later while tarrying in a coffee shop during an unexpected rain storm.

Those thoughts take on extra existential piquancy when they occur during my travels in Europe. 🧑‍🎨🪗

An ellipsis does seem strongly preferable to silent truncation, so presumably we're going to merge this, but Lukas's nits seem valid, and I'd also like my memory refreshed about:

> The new pretty version is not yet in use, IIUC?

**Comment by lrytz** [comment]
I think we can close this in favor of https://github.com/scala/scala/pull/10180, where the nit is also addressed

**Comment by som-snytt** [comment]
I need to add a comment on the (unused) API on the other PR.

**Pull request closed without merging**